Avoid manual `Qk` prefixes in `cbindgen.toml`

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

As suggested in https://github.com/mozilla/cbindgen/issues/1061, we can avoid manually prefixing every object in the C API with `Qk`, by just adding 
```c
typedef PyObject QkPyObject
```
to the header. This will alleviate the problem that we overwrite `PyObject` to `QkPyObject` in our function signatures, which was the reason that we started the manual prefixes.

### Details and comments

The only in-elegant thing is an additional typedef, but this seems the lesser evil than adding manual typedefs for every single object.

The delay unit was renamed since it shouldn't have the `Qk` prefix in the original name.

Cryoris committed:
Avoid manual `Qk` prefixes

mtreinish commented on crates/cext/cbindgen.toml:
```diff
@@ -7,6 +7,7 @@ cpp_compat = true
 after_includes = """
 #ifdef QISKIT_C_PYTHON_INTERFACE
     #include <Python.h>
+    typedef PyObject QkPyObject;
```
What is `QkPyObject`? We don't have our own PyObject type it should be the symbol from Python right?

Cryoris replied:
Exactly -- since we use PyObject in a function signature, cbindgen detects it and adds the generic export prefix, hence it becomes "QkPyObject". To avoid that I suggested cbindgen to add an option to exclude certain objects from prefixing, but that didn't go through. Instead the suggested solution is what this PR does 🙂

Didgety replied:
Did I do something wrong? I see my comment was deleted

Cryoris replied:
@Didgety: I deleted the comment since it seemed off-topic. It didn't explain where `QkPyObject` came from, but instead generally explained the prefix `Qk`.

jakelishman replied:
Got to say I don't love that we do this.  Is there any possibility at all that we could use the `build.rs` script to "undo" the changes to `QkPyObject` instead of leaking out bad tool behaviour to users?  Maybe we could call `cbindgen` in the build script, then rewrite the file to find-and-replace `QkPyObject` back to `PyObject`?

Cryoris committed:
Merge branch 'main' into c-better-prefixing

Cryoris committed:
Merge branch 'main' into c-better-prefixing

Cryoris committed:
Fix sabre naming

Cryoris committed:
Merge branch 'main' into c-better-prefixing

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -1080,7 +1080,7 @@ pub unsafe extern "C" fn qk_circuit_to_python(circuit: *mut CircuitData) -> *mut
 ///
 /// Units for circuit delays.
 #[repr(u8)]
-pub enum QkDelayUnit {
+pub enum CDelayUnit {
```
This is a very sensible solution to the `cbindgen` naming conflicts that I'm surprised took us this long to come up with, given we already do it for things exposed to Python space...

jakelishman reviewed (commented):
Honestly, personally I think I'd rather have the manual list of each class if the alternative is that we put out nonsense like `QkPyObject` and have to "correct" it with typedefs.  But if we're able to undo those silly things, I'm not so against doing this.

Cryoris commented:
> Maybe we could call cbindgen in the build script, then rewrite the file to find-and-replace QkPyObject back to PyObject?

That's a good idea! I'm also not very fond of this artificial renamings (neither the `QkPyObject`, not the `Qk<Anything>`), if we can do it via script that'd tick both boxes

Cryoris committed:
rewrite QkPyObject->PyObject manually

Cryoris committed:
Merge branch 'main' into c-better-prefixing

Cryoris committed:
Merge branch 'c-better-prefixing' of github.com:Cryoris/qiskit-terra into c-better-prefixing

Cryoris commented:
I changed to manual renaming in 75d6d34 -- it would be nicer if we don't have to write the file twice (e.g. by doing this renaming while cbindgen is writing) but this also won't be a performance bottleneck in the compilation so maybe it's fine 😄

Cryoris committed:
Merge branch 'main' into c-better-prefixing

jakelishman commented on crates/cext/build.rs:
```diff
@@ -34,7 +43,29 @@ fn generate_qiskit_header() {
+    let final_file = File::create(&path).expect("The qiskit.h path should exist.");
+    let mut final_buffer = BufWriter::new(final_file);
+    let tmp_file = File::open(&tmp_header).expect("The tmp path should exist.");
+    let tmp_buffer = BufReader::new(tmp_file);
```
Can we simplify this a little bit by writing out the `cbindgen` result into a `String` in memory?  We should be able to use `Bindings::write` to write at least into a `Vec<u8>` and then go `String::from_utf8` or something.

Cryoris replied:
I was unsure about the size of the file here (that also goes to your second comment) and tried to write it such that we don't have to store the full file -- but maybe that concern is unfounded 😄

jakelishman replied:
The file is a small handful of kB at most which is many orders of magnitude smaller than a problem, but either way, the C compiler would have to be able to load it into memory anyway, so it shouldn't be a problem.

jakelishman commented on crates/cext/build.rs:
```diff
@@ -34,7 +43,29 @@ fn generate_qiskit_header() {
+            line = line.replace(target, replacement);
+        }
+        writeln!(final_buffer, "{}", line).expect("Failed writing to qiskit.h");
+    }
```
Do we need to go line by line?  Is `tmp_buffer.replace(target, replacement)` enough by itself?

Cryoris committed:
Use tmp buffer

and assume we can store the full header file as String (which should be fine as long as it doesn't get humongous)

jakelishman reviewed (dismissed):
This seems a bit nicer to maintain for us, thanks.

jakelishman commented:
This failed lint, but is ready to go once fixed.

Cryoris committed:
fix clippy and check all bytes were written

Cryoris commented:
Fixed! This uncovered that the build process could've passed without the full header being written. I added a check for this now too.

jakelishman reviewed (approved):
Thanks - let's try this for a while and see how it goes.
